CASSANDRA-18070 trunk: Add SELECT_MASKED permission

adelapena opened the pull request:
Add new SELECT_MASKED permission. It allows to run SELECT queries filtering the clear values of masked columns.

Superusers have it by default, whereas regular users don't have it by default.

Users without UNMASK nor SELECT_MASKED permission won't be authorised to use masked columns on the WHERE part of a SELECT query.

adelapena committed:
Add masking functions to column metadata

Allows to attach the native masking functions added by CASSANDRA-17941 to the definitions of
table columns in the schema, as defined by CEP-20. The functions masking a column can be
specified on CREATE TABLE queries, right at the end of the column definition. The mask of a
column can also be changed or dropped with an ALTER TABLE query. Once a column is masked,
SELECT queries will always return the masked value of the column. That masking is done on the
coordinator, at the end of the query execution. Thus, masking won't affect any filters or
ordering, which would be based on the clear values of the masked columns. Column masks are
stored on the system table system_schema.column_masks.

patch by Andrés de la Peña; reviewed by Benjamin Lerer and Berenguer Blasi for CASSANDRA-18068

bereng commented on src/java/org/apache/cassandra/cql3/statements/SelectStatement.java:
```diff
@@ -238,7 +239,21 @@ public void authorize(ClientState state) throws InvalidRequestException, Unautho
+        {
+            List<ColumnMetadata> queriedMaskedColumns = table.columns()
+                                                             .stream()
+                                                             .filter(ColumnMetadata::isMasked)
```
Would it be better to filter `isRestricted` first?

adelapena replied:
[`StatementRestrictions#isRestricted`](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java#L341-L346) involves building a collection [in most cases](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java#L60) and then running a `List#contains`. In contrast, {{ColumnMetadata#isMasked}} just returns a boolean. So I think `isMasked` is faster than is `isRestricted`.

bereng replied:
sgtm

bereng commented on test/unit/org/apache/cassandra/cql3/functions/masking/QueryMaskedPermissionTest.java:
```diff
@@ -0,0 +1,391 @@
+        Object[] maskedRow = row(0, 7, 7, 7);
+
+        // test queries with default permissions (no UNMASK nor SELECT_MASKED)
+        assertWithAuthSpin(() -> {
```
If my eyes aren't fooling me aren't we just repeating the same queries over and over again? Did you do this intentionally or would you consider some type of looping to avoid the repeats?

adelapena replied:
I think that the queries tested of every block are different to each other. And the blocks are different to each other on whether they return a masked or unmasked row, or an authorization failure. The only blocks that are repeated are the ones for only `UNMASK` permission and for both `UNMASK` and `SELECT_MASKED` permissions, and the ones for defaults and manually removed permissions. I can put those pairs on blocks on separate methods to save some duplication, although I'm not sure that actually improves readability. 

As for the differences test to test, they might look similar but they aren't identical, since different schemas allow different queries.

adelapena committed:
Add UNMASK permission

The new permission allows to see the clear data of columns with an attached mask.
Superusers have it by default, whereas regular users don't have it.

patch by Andrés de la Peña; reviewed by Benjamin Lerer and Berenguer Blasi for CASSANDRA-18069

bereng commented on doc/modules/cassandra/pages/cql/cql_singlefile.adoc:
```diff
@@ -2427,6 +2428,12 @@ wildcard pattern | | |
+
+|`SELECT_MASKED` | `KEYSPACE` | `SELECT` restricting masked columns on any table in specified keyspace | | |
+
+|`SELECT_MASKED` | `TABLE` | `SELECT` restricting masked columns on specified table | | |
```
Wording?

adelapena committed:
Add SELECT_MASKED permission

The new permission allows to run SELECT queries restricting the clear values of masked columns.
Superusers have it by default, whereas regular users don't have it.

patch by Andrés de la Peña; reviewed by Benjamin Lerer and Berenguer Blasi for CASSANDRA-18070

adelapena committed:
DO NOT MERGE - CircleCI config

pmcfadin commented:
Closing as [CASSANDRA-18070](https://issues.apache.org/jira/browse/CASSANDRA-18070) has been resolved as Fixed — if this PR contains changes not captured in the merged fix, please reopen.
